source/extensions/geoip_providers/maxmind/geoip_provider.cc

```diff
@@ -0,0 +1,279 @@
+  anon_proxy_header_ = !geo_headers_to_add.anon_proxy().empty()
+                           ? absl::make_optional(geo_headers_to_add.anon_proxy())
+                           : absl::nullopt;
+  if (!city_db_path_ && !isp_db_path_ && !anon_db_path_) {
+    throw EnvoyException("At least one geolocation database path needs to be configured: "
+                         "city_db_path, isp_db_path or anon_db_path");
+  }
+  if (city_db_path_) {
+    registerGeoDbStats("city_db");
+  }
+  if (isp_db_path_) {
+    registerGeoDbStats("isp_db");
+  }
+  if (anon_db_path_) {
+    registerGeoDbStats("anon_db");
+  }
+};
+
+void GeoipProviderConfig::registerGeoDbStats(const std::string& db_type) {
+  stat_name_set_->rememberBuiltin(absl::StrCat(db_type, ".total"));
+  stat_name_set_->rememberBuiltin(absl::StrCat(db_type, ".hit"));
+  stat_name_set_->rememberBuiltin(absl::StrCat(db_type, ".lookup_error"));
+}
+
+bool GeoipProviderConfig::isLookupEnabledForHeader(const absl::optional<std::string>& header) {
+  return (header && !header.value().empty());
+}
+
+void GeoipProviderConfig::incCounter(Stats::StatName name) {
+  Stats::SymbolTable::StoragePtr storage = scope_.symbolTable().join({stats_prefix_, name});
+  scope_.counterFromStatName(Stats::StatName(storage.get())).inc();
+}
+
+GeoipProvider::~GeoipProvider() {
+  ENVOY_LOG(debug, "Shutting down Maxmind geolocation provider");
+  // todo(nezdolik) consider switching to atomic singleton for db objects
+  if (city_db_) {
+    MMDB_close(city_db_.get());
+  }
+  if (isp_db_) {
+    MMDB_close(isp_db_.get());
+  }
+  if (anon_db_) {
+    MMDB_close(anon_db_.get());
+  }
+}
+
+MaxmindDbPtr GeoipProvider::initMaxMindDb(const absl::optional<std::string>& db_path) {
+  if (db_path) {
+    MMDB_s maxmind_db;
+    int result_code = MMDB_open(db_path.value().c_str(), MMDB_MODE_MMAP, &maxmind_db);
+    RELEASE_ASSERT(MMDB_SUCCESS == result_code,
+                   fmt::format("Unable to open Maxmind database file {}. Error {}", db_path.value(),
+                               std::string(MMDB_strerror(result_code))));
+    return std::make_unique<MMDB_s>(maxmind_db);
+  } else {
+    ENVOY_LOG(debug, "Geolocation database path is empty, skipping database creation");
+    return nullptr;
+  }
+}
+
+void GeoipProvider::lookup(Geolocation::LookupRequest&& request,
+                           Geolocation::LookupGeoHeadersCallback&& cb) const {
+  auto& remote_address = request.remoteAddress();
+  auto lookup_result = absl::flat_hash_map<std::string, std::string>{};
+  std::cerr << "***Lookup for ip: " << remote_address->ip()->addressAsString() << std::endl;
+  lookupInCityDb(remote_address, lookup_result);
+  lookupInAsnDb(remote_address, lookup_result);
+  lookupInAnonDb(remote_address, lookup_result);
+  cb(std::move(lookup_result));
+}
+
+// todo: document En as default and only language used for lookups. Document that region is selcted
+// as least specific subdivision
+void GeoipProvider::lookupInCityDb(
```

Comment:
there is code duplication across lookupInCityDb(), lookupInAsnDb() and lookupInAnonDb(), i have tried to merge those into single method but resulting code was not pretty due to calls to C variadic function (in maxmind api) here: https://github.com/envoyproxy/envoy/pull/28490/files#diff-d21c11080878cdf65727845da9fa3846d138efb810faeb12e8d16006713446deR259